lib/Flow/Operation.php

```diff
@@ -0,0 +1,244 @@
+use OCP\IL10N;
+use OCP\IURLGenerator;
+use OCP\IUser;
+use OCP\IUserSession;
+use OCP\WorkflowEngine\EntityContext\IDisplayText;
+use OCP\WorkflowEngine\IManager as FlowManager;
+use OCP\WorkflowEngine\IOperation;
+use OCP\WorkflowEngine\IRuleMatcher;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use UnexpectedValueException;
+
+class Operation implements IOperation {
+
+	/** @var int[] */
+	public const MESSAGE_MODES = [
+		'NO_MENTION' => 1,
+		'SELF_MENTION' => 2,
+		'ROOM_MENTION' => 3,
+	];
+
+	/** @var IL10N */
+	private $l;
+	/** @var IURLGenerator */
+	private $urlGenerator;
+	/** @var TalkManager */
+	private $talkManager;
+	/** @var IUserSession */
+	private $session;
+	/** @var ChatManager */
+	private $chatManager;
+
+	public function __construct(
+		IL10N $l,
+		IURLGenerator $urlGenerator,
+		TalkManager $talkManager,
+		IUserSession $session,
+		ChatManager $chatManager
+	) {
+		$this->l = $l;
+		$this->urlGenerator = $urlGenerator;
+		$this->talkManager = $talkManager;
+		$this->session = $session;
+		$this->chatManager = $chatManager;
+	}
+
+	public static function register(EventDispatcherInterface $dispatcher): void {
+		$dispatcher->addListener(FlowManager::EVENT_NAME_REG_OPERATION, function (GenericEvent $event) {
+			$operation = \OC::$server->query(Operation::class);
+			$event->getSubject()->registerOperation($operation);
+			OC_Util::addScript('spreed', 'flow');
+		});
+	}
+
+	public function getDisplayName(): string {
+		return $this->l->t('Write to conversation');
+	}
+
+	public function getDescription(): string {
+		return $this->l->t('Writes event information into a conversation of your choice');
+	}
+
+	public function getIcon(): string {
+		return $this->urlGenerator->imagePath('spreed', 'app.svg');
+	}
+
+	public function isAvailableForScope(int $scope): bool {
+		return $scope === FlowManager::SCOPE_USER;
+	}
+
+	/**
+	 * Validates whether a configured workflow rule is valid. If it is not,
+	 * an `\UnexpectedValueException` is supposed to be thrown.
+	 *
+	 * @throws UnexpectedValueException
+	 * @since 9.1
+	 */
+	public function validateOperation(string $name, array $checks, string $operation): void {
+		list($mode, $token) = $this->parseOperationConfig($operation);
+		$this->validateOperationConfig($mode, $token, $this->getUser()->getUID());
+	}
+
+	public function onEvent(string $eventName, Event $event, IRuleMatcher $ruleMatcher): void {
+		$flows = $ruleMatcher->getFlows(false);
+		foreach ($flows as $flow) {
+			try {
+				list($mode, $token) = $this->parseOperationConfig($flow['operation']);
+				$uid = $flow['scope_actor_id'];
+				$this->validateOperationConfig($mode, $token, $uid);
+
+				$entity = $ruleMatcher->getEntity();
+
+				$message = $eventName;
+				if($entity instanceof IDisplayText) {
+					$message = $entity->getDisplayText(3);
+					if($message === '') {
+						continue;
+					}
+				}
```

Comment:
Adding a link if IUrl is implemented would be nice as well

Reply:
Yeah, makes sense. Then, only when some other text is already given, since a single url alone is too strange in a conversation